### PR TITLE
added kwargs option to MH callback; added model summary; added accept…

### PR DESCRIPTION
**Pull request by DzvinkaYarish**
…ance ratio logging

### CODE REVIEW

File: deep_tempering/callbacks.py [hunk 1]
```diff
@@ -31,7 +31,8 @@ def configure_callbacks(callbacks,
                         mode=ModeKeys.TRAIN,
                         exchange_data=None,
                         swap_step=None,
-                        burn_in=None):
+                        burn_in=None,
+                        **kwargs):
```
Comment (vlad-user) [inline deep_tempering/callbacks.py]: 1. The `kwargs` should be passed directly to the `ExchangeCallback`.  It is not good idea to carry kwargs all over the place (passing to each func - `fit()`. By passing kwargs to the `ExchangeCallback` user chooses desired behaviour. For example, if you want to define a callback with non-default value of coefficient, you define:
```python
clbk = MetropolisExchangeCallback(exchange_data, swap_step, burn_in, coeff=2.)
model.fit(..., callbacks=[clbk]
```
Otherwise, the default behaviour is used. This way there is no need to modify `fit()` arguments for each new variable we wish to use.

You could also create your own callbacks (by subclassing `BaseExchangeCallback`), it doesn't have to be a part of this package, especially on the stage of the prototyping. If at the end we will see that the callbacks work well and we could use them as default behaviour we could add them here as well. 

2. `swap_step` and `burn_in` is also better to remove (but let's keep it as is for now).

**Review by vlad-user — CHANGES_REQUESTED** [review]
Good. Please read my comments.
+ Have you ran the `pytest`?

File: deep_tempering/callbacks.py [hunk 2]
```diff
@@ -485,8 +482,10 @@ def exchange(self, *args, **kwargs):
 
 class MetropolisExchangeCallback(BaseExchangeCallback):
   """Exchanges of hyperparameters based on Metropolis acceptance criteria."""
-  def __init__(self, exchange_data, swap_step, burn_in=None):
+  def __init__(self, exchange_data, swap_step=1, burn_in=1, coeff=1., hp_to_swap=None):
```
Comment (vlad-user) [inline deep_tempering/callbacks.py]: `hp_to_swap` should not be a part of the default exchange procedure. The default exchange is as is. If user wants to create other exchange procedure it needs to write his own callback.

File: deep_tempering/callbacks.py [hunk 3]
```diff
@@ -497,14 +496,17 @@ def exchange(self, **kwargs):
     """
     # pick random hyperparameter to exchange
     hp = self.ordered_hyperparams
-    hpname = kwargs.get('hpname', random.choice(list(hp.keys())))
+    hpname = kwargs.get('hpname', None)
+    if not hpname:
+      hpname = self.hpname if self.hpname else random.choice(list(hp.keys()))
```
Comment (vlad-user) [inline deep_tempering/callbacks.py]: same here.

File: deep_tempering/callbacks.py [hunk 4]
```diff
@@ -519,21 +521,26 @@ def exchange(self, **kwargs):
...
 
     if np.random.uniform() < proba:
       swaped = 1
-      self.model.hpspace.swap_between(i, j, hpname)
+      self.model.hpspace.swap_between(replicas_ids[i], replicas_ids[j], hpname)
     else:
       swaped = 0
 
+    if getattr(self, 'exchange_logs', None):
+      accpt_ratio = (self.exchange_logs['swaped'].count(1) + swaped) / (len(self.exchange_logs['proba']) + 1)
```
Comment (vlad-user) [inline deep_tempering/callbacks.py]: please make sure that all lines are below conventional 80 characters.

File: deep_tempering/callbacks.py [hunk 5]
```diff
@@ -519,21 +521,26 @@ def exchange(self, **kwargs):
...
+    else:
+      accpt_ratio = swaped
+
     super().log_exchange_metrics(losses, proba=proba, hpname=hpname,
-                                 swaped=swaped)
+                                 swaped=swaped, accept_ratio=accpt_ratio, delta=delta, exchange_pair=[replicas_ids[i], replicas_ids[j]])
```
Comment (vlad-user) [inline deep_tempering/callbacks.py]: same here. please make sure that all lines are below conventional 80 characters.

File: deep_tempering/callbacks.py [hunk 6]
```diff
@@ -519,21 +521,26 @@ def exchange(self, **kwargs):
...
 
 
 class MonitorOptimalModelCallback(tf.keras.callbacks.Callback):
-  """Monitors optimal keras' model.
+  """Monitors optimal keras' model].
```
Comment (vlad-user) [inline deep_tempering/callbacks.py]: typo

File: deep_tempering/training_utils.py [hunk 7]
```diff
@@ -150,7 +150,7 @@ def prepare_feed_tensors_and_values(self, training=True):
     for i in range(n_replicas):
       for hpname in hpnames:
         if not training and 'dropout_rate' in hpname:
-          value = 0.
+          value = 1.
```
Comment (vlad-user) [inline deep_tempering/training_utils.py]: if this is not training the dropout rate should be 0. Why did you change it to 1?

**Review by vlad-user — CHANGES_REQUESTED** [review]
Please look at my comments. 
Have you run the tests?

File: deep_tempering/training_utils.py (hunk 7 above)
Comment (DzvinkaYarish) [reply deep_tempering/training_utils.py]: Older version of tf uses keep probability instead of dropout rate. But now this issue is managed in my custom model directly

**Comment by DzvinkaYarish** [comment]
> Please look at my comments.
> Have you run the tests?

Yes, the tests have all passed

File: deep_tempering/callbacks.py [hunk 8]
```diff
@@ -519,17 +519,28 @@ def exchange(self, **kwargs):
...
-      self.model.hpspace.swap_between(i, j, hpname)
+      self.model.hpspace.swap_between(replicas_ids[i], replicas_ids[j], hpname)
     else:
       swaped = 0
 
-    super().log_exchange_metrics(losses, proba=proba, hpname=hpname,
-                                 swaped=swaped)
+    if getattr(self, 'exchange_logs', None):
+      accpt_ratio = (self.exchange_logs['swaped'].count(1) + swaped) / \
+                    (len(self.exchange_logs['proba']) + 1)
```
Comment (vlad-user) [inline deep_tempering/callbacks.py]: Please use PEP8 conventions. Don't use `\` for line continuation. If you can't fit to line you could create temporary variables. Eg:
```python
if getattr(self, 'exchange_logs', None):
  n_attempts = len(self.exchange_logs['proba']) + 1
  n_swaps = self.exchange_logs['swaped'].count(1) + swaped
  accpt_ratio = n_swaps / n_attempts
```

**Review by vlad-user — CHANGES_REQUESTED** [review]
Please see my last comment and merge after that.
Good job.
